core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -701,22 +707,29 @@ class KafkaApis(val requestChannel: RequestChannel,
     var (authorizedTopics, unauthorizedTopics) =
       topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
 
+    var uncreatableTopics = Set[String]()
+      
     if (authorizedTopics.nonEmpty) {
       val nonExistingTopics = metadataCache.getNonExistingTopics(authorizedTopics)
       if (config.autoCreateTopicsEnable && nonExistingTopics.nonEmpty) {
         authorizer.foreach { az =>
           if (!az.authorize(request.session, Create, Resource.ClusterResource)) {
             authorizedTopics --= nonExistingTopics
-            unauthorizedTopics ++= nonExistingTopics
+            uncreatableTopics ++= nonExistingTopics
           }
         }
       }
     }
 
-    val unauthorizedTopicMetadata = unauthorizedTopics.map(topic =>
+    val uncreatableTopicMetadata = uncreatableTopics.map(topic =>
       new MetadataResponse.TopicMetadata(Errors.TOPIC_AUTHORIZATION_FAILED, topic, common.Topic.isInternal(topic),
         java.util.Collections.emptyList()))
 
+    // do not disclose the existence of the unauthorized topic    
+    val unauthorizedTopicMetadata = unauthorizedTopics.map(topic =>
```

Comment:
I think we need to handle the all topics case specially:

```scala
 val topics =
       // Handle old metadata request logic. Version 0 has no way to specify "no topics".
       if (requestVersion == 0) {
         if (metadataRequest.topics() == null || metadataRequest.topics().isEmpty)
           metadataCache.getAllTopics()
         else
           metadataRequest.topics.asScala.toSet
       } else {
         if (metadataRequest.isAllTopics)
           metadataCache.getAllTopics()
         else
           metadataRequest.topics.asScala.toSet
       }
```

For that case, we want to filter the topics that the user doesn't have DESCRIBE access to.